expression/builtin_time.go

```diff
@@ -332,83 +339,319 @@ type timeDiffFunctionClass struct {
+	d, isNull, err = calculateTimeDiff(lhs, rhs, sc)
+	return d, isNull, errors.Trace(err)
+}
+
+type builtinDurationStringTimeDiffSig struct {
+	baseDurationBuiltinFunc
 }
 
-// eval evals a builtinTimeDiffSig.
+// evalDuration evals a builtinDurationStringTimeDiffSig.
 // See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_timediff
-func (b *builtinTimeDiffSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
+func (b *builtinDurationStringTimeDiffSig) evalDuration(row []types.Datum) (d types.Duration, isNull bool, err error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	lhs, isNull, err := b.args[0].EvalDuration(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
 	}
-	if args[0].IsNull() || args[1].IsNull() {
-		return
+
+	rhsStr, isNull, err := b.args[1].EvalString(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
 	}
 
-	sc := b.ctx.GetSessionVars().StmtCtx
-	fsp := int(math.Max(float64(args[0].Frac()), float64(args[1].Frac())))
-	t0, err := b.convertArgToTime(sc, args[0], fsp)
+	rhs, _, isDuration, err := convertStringToDuration(rhsStr, b.tp.Decimal, sc)
 	if err != nil {
-		return d, errors.Trace(err)
+		return d, true, errors.Trace(err)
+	}
+
+	if !isDuration {
+		return d, true, nil
+	}
+
+	d, isNull, err = calculateDurationTimeDiff(lhs, rhs, sc)
+	return d, isNull, errors.Trace(err)
+}
+
+type builtinStringDurationTimeDiffSig struct {
+	baseDurationBuiltinFunc
+}
+
+// evalDuration evals a builtinStringDurationTimeDiffSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_timediff
+func (b *builtinStringDurationTimeDiffSig) evalDuration(row []types.Datum) (d types.Duration, isNull bool, err error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	lhsStr, isNull, err := b.args[0].EvalString(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
+	}
+
+	rhs, isNull, err := b.args[1].EvalDuration(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
 	}
-	t1, err := b.convertArgToTime(sc, args[1], fsp)
+
+	lhs, _, isDuration, err := convertStringToDuration(lhsStr, b.tp.Decimal, sc)
 	if err != nil {
-		return d, errors.Trace(err)
+		return d, true, errors.Trace(err)
 	}
-	if (types.IsTemporalWithDate(t0.Type) &&
-		t1.Type == mysql.TypeDuration) ||
-		(types.IsTemporalWithDate(t1.Type) &&
-			t0.Type == mysql.TypeDuration) {
-		return d, nil // Incompatible types, return NULL
+	if !isDuration {
+		return d, true, nil
 	}
 
-	t := t0.Sub(&t1)
-	ret, truncated := types.TruncateOverflowMySQLTime(t.Duration)
-	if truncated {
-		err = types.ErrTruncatedWrongVal.GenByArgs("time", t.String())
+	d, isNull, err = calculateDurationTimeDiff(lhs, rhs, sc)
+	return d, isNull, errors.Trace(err)
+}
+
+func calculateTimeDiff(t0, t1 types.Time, sc *variable.StatementContext) (d types.Duration, isNull bool, err error) {
```

Comment:
- move `sc *variable.StatementContext` to the first parameter 
- `s/t0/lhs/` and `s/t1/rhs/`